src/backend/distributed/executor/distributed_intermediate_results.c

```diff
@@ -59,28 +59,6 @@ typedef struct PartitioningTupleDest
 } PartitioningTupleDest;
 
 
-/*
- * NodePair contains the source and destination node in a NodeToNodeFragmentsTransfer.
- * It is a separate struct to use it as a key in a hash table.
- */
-typedef struct NodePair
-{
-	uint32 sourceNodeId;
-	uint32 targetNodeId;
-} NodePair;
-
-
-/*
- * NodeToNodeFragmentsTransfer contains all fragments that need to be fetched from
- * the source node to the destination node in the NodePair.
- */
-typedef struct NodeToNodeFragmentsTransfer
-{
-	NodePair nodes;
-	List *fragmentList;
-} NodeToNodeFragmentsTransfer;
-
-
 /* forward declarations of local functions */
 static List * WrapTasksForPartitioning(const char *resultIdPrefix,
```

Comment:
Why don't we change the parameters of `worker_partition_query_result ` in `WrapTasksForPartitioning`?

I think I'm a bit confused about the `skipNullPartitionColumnValue ` and `generateEmptyResults ` flags. Are they needed for SELECT queries but not for `INSERT .. SELECT`? Can you clarify a bit for me to follow easier? Thanks!

Reply:
> skipNullPartitionColumnValue 

INSERT..SELECT fails when there are NULL values:

> ERROR:  the partition column value cannot be NULL
> CONTEXT:  while executing command on localhost:1402

That's because we currently do not allow NULL values in the partition columns, and FindShardInterval does not work for NULL values.

We do allow NULL values as the input to re-partition joins, but they will never match the join condition, so we might as well skip them.

We will have to revisit this logic if we re-partition for other purposes than joins (e.g. group by). The solution would be to implement shard pruning for NULL, and maybe generally allow NULL as a partition column value.

> generateEmptyResults 

We do not prune away fetch tasks, so fetching an empty file would fail. I tried making COPY .. TO STDOUT WITH (format 'result') tolerate non-existent files, but it got quite complicated because that command does not know the actual formatting of the file (binary vs. text), which then needs to be handled somewhere downstream. Instead, we just generate empty files and fetch them like we always did. One day we'll have pruning like we do in insert..select and the need for this goes away.

Reply:
(will add some comments in multi_physical_planner.c)